### PR TITLE
Default web templates to net5.0

### PR DESCRIPTION
**Pull request by ViktorHofer**
Updating the web templates to default to net5.0 instead of netcoreapp5.0.
Contributes towards https://github.com/dotnet/sdk/issues/10756.

cc @anurse @Pilchie 

This is required for preview2 which closes down EOW.

### CODE REVIEW

**Comment by mkArtakMSFT** [comment]
@pranavkm can you please review this early in the week (Monday, Tuesday)? We need to get this in early in the week.

**Review by pranavkm — COMMENTED** [review]
The template change looks alright to me, but I'd prefer if someone from @dotnet/aspnet-build signs off on the other changes in the PR

**Comment by ViktorHofer** [comment]
I don't understand why builds are failing with requiring a newer msbuild even though I didn't update the global.json's sdk value. Just updating the dotnet value should only impact which version is bootstrapped but not which minimum version is required by desktop msbuild...

File: eng/Workarounds.targets [hunk 1]
```diff
@@ -40,17 +40,22 @@
...
+    <KnownAppHostPackOrFrameworkReferenceTfm Condition="'$(KnownAppHostPackOrFrameworkReferenceTfm)' == 'net5.0'">netcoreapp5.0</KnownAppHostPackOrFrameworkReferenceTfm>
+  </PropertyGroup>
+
   <ItemGroup>
-    <!-- Workaround when there is no vNext SDK available, copy known apphost/framework reference info from 3.0 -->
+    <!-- Workaround when there is no vNext SDK available, copy known apphost/framework reference info from 3.1 -->
     <KnownAppHostPack
-      Include="@(KnownAppHostPack->WithMetadataValue('TargetFramework', 'netcoreapp3.0'))"
+      Include="@(KnownAppHostPack->WithMetadataValue('TargetFramework', 'netcoreapp3.1'))"
       TargetFramework="$(DefaultNetCoreTargetFramework)"
```
Comment (dougbu) [inline eng/Workarounds.targets]: Suspicious as-is
```suggestion
      TargetFramework="$(KnownAppHostPackOrFrameworkReferenceTfm)"
```

File: eng/Workarounds.targets [hunk 2]
```diff
@@ -40,17 +40,22 @@
...
-      Condition="@(KnownAppHostPack->Count()) != '0' AND !(@(KnownAppHostPack->AnyHaveMetadataValue('TargetFramework', '$(DefaultNetCoreTargetFramework)')))"
+      Condition="@(KnownAppHostPack->Count()) != '0' AND !(@(KnownAppHostPack->AnyHaveMetadataValue('TargetFramework', '$(KnownAppHostPackOrFrameworkReferenceTfm)')))"
       />
     <KnownFrameworkReference
-      Include="@(KnownFrameworkReference->WithMetadataValue('TargetFramework', 'netcoreapp3.0'))"
+      Include="@(KnownFrameworkReference->WithMetadataValue('TargetFramework', 'netcoreapp3.1'))"
       TargetFramework="$(DefaultNetCoreTargetFramework)"
```
Comment (dougbu) [inline eng/Workarounds.targets]: Suspicious as-is
```suggestion
      TargetFramework="$(KnownAppHostPackOrFrameworkReferenceTfm)"
```

File: global.json [hunk 3]
```diff
@@ -3,7 +3,7 @@
     "version": "5.0.100-preview.2.20120.3"
```
Comment (dougbu) [inline global.json]:
```suggestion
    "version": "5.0.100-preview.2.20163.4"
```

File: global.json [hunk 4]
```diff
@@ -1,9 +1,11 @@
 {
   "sdk": {
-    "version": "5.0.100-preview.2.20120.3"
+    "version": "5.0.100-preview.2.20120.3",
```
Comment (dougbu) [inline global.json]:
```suggestion
    "version": "5.0.100-preview.2.20163.4",
```

File: src/ProjectTemplates/build.cmd [hunk 5]
```diff
@@ -1,3 +1,3 @@
 @ECHO OFF
 SET RepoRoot=%~dp0..\..
-%RepoRoot%\build.cmd -projects %~dp0*\*.*proj "/p:EnforceE2ETestPrerequisites=true" %*
+%RepoRoot%\build.cmd -projects %~dp0*\*.*proj -ForceCoreMsbuild "/p:EnforceE2ETestPrerequisites=true" %*
```
Comment (dougbu) [inline src/ProjectTemplates/build.cmd]: Why is `-ForceCoreMsbuild` needed?

File: global.json (hunk 4 above)
Comment (ViktorHofer) [reply global.json]: I don't want to update the necessary SDK as desktop msbuild should still work with 16.4. I just want to update the bootstrapped dotnet SDK.

**Comment by ViktorHofer** [comment]
@dougbu thanks for reviewing this. It would great if you can help me with https://github.com/dotnet/aspnetcore/pull/19860#issuecomment-599550849.

File: global.json (hunk 4 above)
Comment (dougbu) [reply global.json]: We tend to move these values in sync and expect devs to at least run `./restore` before using VS

Comment (dsplaisted) [reply global.json]: I think you will need to switch to MSBuild 16.5.  The updates to dotnet/sdk to support net5.0 as the TargetFramework [use new intrinsic MSBuild functions](https://github.com/dotnet/sdk/blob/983c8115e6341287a5866ed7be05c663c7e89614/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.TargetFrameworkInference.targets#L72-L75) which were [added in MSBuild 16.5](https://github.com/microsoft/msbuild/pull/4911).

**Review by dougbu — COMMENTED** [review]
> Just updating the dotnet value should only impact which version is bootstrapped but not which minimum version is required by desktop msbuild...

But the CI build will use the bootstrapped SDK and that requires a newer version of VS.

What surprises me is the lack of 16.5 on the Windows agents. @MattGal shouldn't the BuildPool.Server.Amd64.VS2019.BT.Open agents be more up-to-date?

File: global.json (hunk 4 above)
Comment (dougbu) [reply global.json]: Won't changing the required VS or MSBuild version to 16.5 in this file just cause Windows builds to fail faster on these agents? From the errors, they only have 16.4 installed.

**Comment by MattGal** [comment]
> > Just updating the dotnet value should only impact which version is bootstrapped but not which minimum version is required by desktop msbuild...
> 
> But the CI build will use the bootstrapped SDK and that requires a newer version of VS.
> 
> What surprises me is the lack of 16.5 on the Windows agents. @MattGal shouldn't the BuildPool.Server.Amd64.VS2019.BT.Open agents be more up-to-date?

Latest public version of VS 2019 is 16.4.6.  The VS 2019 BT queues currently have 16.4.3.  Since that's not a preview queue, until some time after that's publicly released I don't think they should be more up-to-date as that would require running a Preview SKU.

**Comment by ViktorHofer** [comment]
Thanks Matt for the infos. I pinged @jaredpar who will update the RoslynTools.MSBuild package (which contains an xcopyable verison of MSBuild) to 16.5.0 preview. This will allow us to require a newer vs version in global.json which will fix the current failures.

**Comment by dougbu** [comment]
I _think_ another option may be to change the .azure/pipelines/jobs/default-build.yaml file to switch the pools to BuildPool.Server.Amd64.VS2019.BT.Pre.Open and BuildPool.Server.Amd64.VS2019.BT.Pre. (Did I get those names right @MattGal?)

**Comment by MattGal** [comment]
> I _think_ another option may be to change the .azure/pipelines/jobs/default-build.yaml file to switch the pools to BuildPool.Server.Amd64.VS2019.BT.Pre.Open and BuildPool.Server.Amd64.VS2019.BT.Pre. (Did I get those names right @MattGal?)

Neither of those queues exist.  Basically we only started having these before we'd figured out how to properly automatically activate VS on our machines, and the time-bomb prompts were causing some builds to hang.  

You could try these two:
```
buildpool.server.amd64.vs2019.pre
buildpool.server.amd64.vs2019.pre.open
```

But if it's at all possible (we recognize that it isn't always) you should bootstrap such pre-release dependencies as part of your builds as Viktor mentions.  There are other benefits to this as well, such as making OSS contributors able to build and run without having to install prerelease VS themselves.

**Comment by ViktorHofer** [comment]
Unfortunately that didn't solve the issue as the RoslynTools.MSBuild package doesn't contain VC++ assets. @MattGal 16.5 was released yesterday, when will the queues be updated? Regarding the Pre queues, I don't see a BT one for pre.

**Comment by MattGal** [comment]
> Unfortunately that didn't solve the issue as the RoslynTools.MSBuild package doesn't contain VC++ assets. @MattGal 16.5 was released yesterday, when will the queues be updated? Regarding the Pre queues, I don't see a BT one for pre.

I'm checking in with DDFUN folks; there's a PR out that updates these images right now but I'm uncertain whether it picks up the latest VSes.

**Comment by jaredpar** [comment]
@MattGal is this also going to silently roll runtime repositories native toolset forward again? If we're just updating the images in place it seems like we're going to be doing this.

**Comment by MattGal** [comment]
> > Unfortunately that didn't solve the issue as the RoslynTools.MSBuild package doesn't contain VC++ assets. @MattGal 16.5 was released yesterday, when will the queues be updated? Regarding the Pre queues, I don't see a BT one for pre.
> 
I'm checking in with DDFUN folks; there's a PR out that updates these images right now but I'm uncertain whether it picks up the latest VSes.



> @MattGal is this also going to silently roll runtime repositories native toolset forward again? If we're just updating the images in place it seems like we're going to be doing this.

Yes, except for the silent part (there will be a partners mail before such a rollout).  I believe this will also bring the updated CMake that doesn't produce the warnings that were causing issues.  There are options to the buildpool.windows machines; you can also use the (silently upgraded whenever they want as well) hosted-build machines, or bootstrap all your build dependencies at your discretion.

**Comment by jaredpar** [comment]
@MattGal 

>  bootstrap all your build dependencies at your discretion.

This has been discussed endlessly and every time the conclusion is that this is **not** a viable option for native toolsets. 

> Yes, except for the silent part (there will be a partners mail before such a rollout).

Are we testing it? The official builds for dotnet/runtime have been on the floor for almost a week now due to an untested arcade publishing changes. Just now as we're getting past that we're going to get another untested change to the infrastructure. The expectation here, given the history, should be that this will cause further instability. 

We've proposed solutions to this problem, specifically https://github.com/dotnet/arcade/issues/5018, but we're getting no traction on the discussion. 

@jkotas, @jkoritzinsky as FYI that this is likely coming our way.

**Comment by MattGal** [comment]
>Are we testing it? The official builds for dotnet/runtime have been on the floor for almost a week now due to an untested arcade publishing changes.

Yes, we can absolutely run a set of your builds through the staging environment before this rolls out.  I'll coordinate it with @ilyas1974 once the changes are made and in (hopefully tomorrow)

**Comment by v-joshli** [comment]
Hello,

Casey has completed the VS2019 images. The names are:
Helix-Client-Enterprise-RS5-VS2019BuildTools-03-17-2020
Helix-Client-Enterprise-RS5-VS2019Enterprise-03-17-2020
Helix-Server-DataCenter-RS5-VS2019BuildTools-03-17-2020
Helix-Server-Datacenter-RS5-VS2019Enterprise-03-17-2020

Thanks,
Josh

**Comment by ViktorHofer** [comment]
Hey now I'm confused, are those images only for helix? If yes, how do they relate to this discussion?

**Comment by MattGal** [comment]
> Hey now I'm confused, are those images only for helix? If yes, how do they relate to this discussion?

He may have given you more detail than necessary.  Those images underpin the BuildPool.*.VS2019.* Helix queues; if you want 16.5 GA MSBuild on the box, they are very relevant to this discussion albeit excessive detail for the aspnetcore repo.

**Comment by dougbu** [comment]
@ViktorHofer I suspect we need to change many more files to get the templates working correctly. At least the template.json files (where the TFM is hard-coded) should be updated.

More likely, should update `$(DefaultNetCoreTargetFramework)` _everywhere_ to avoid problems like
```
F:\workspace\_work\1\s\src\ProjectTemplates\test\ProjectTemplates.Tests.csproj : error NU1201: Project Microsoft.AspNetCore.Server.IntegrationTesting is not compatible with net50 (.NETFramework,Version=v5.0). Project Microsoft.AspNetCore.Server.IntegrationTesting supports: netcoreapp5.0 (.NETCoreApp,Version=v5.0)
```

**Comment by ViktorHofer** [comment]
> More likely, should update $(DefaultNetCoreTargetFramework) everywhere to avoid problems like

That's not necessary to unblock this PR. The reason why this is failing is because the NuGet version being used doesn't understand `net5.0`. I'm still trying to workaround this.

**Comment by ViktorHofer** [comment]
> I suspect we need to change many more files to get the templates working correctly. At least the template.json files (where the TFM is hard-coded) should be updated.

Thanks, I missed those. Updated.

**Comment by ViktorHofer** [comment]
So I tried different approaches and there's no _easy_ way to not depend on the machine's installed VS version. For preview2 I will do a manual submission of the updated templates, I will send out a mail to Tactics about this and loop in @Pilchie.

Let's keep this open until we have an updated image that uses 16.5 stable.

**Comment by mthalman** [comment]
What's the story on this?  Using `net5.0` for web templates was previously working in SDK version 5.0.100-preview.2.20169.1 but is now failing in 5.0.100-preview.2.20174.8.

**Comment by Pilchie** [comment]
There is a dependency issue that means we can't make this change in Preview 2, so we rolled it back, but we plan to update the default for Preview 3.

**Comment by ViktorHofer** [comment]
I updated the PR to target the temporary pool with VS 16.5.1 installed. I will finish the remaining work over the weekend and we expect to get the non-temporary build pools for both public and internal by next Wednesday. Additionally, I'll retarget this PR to preview3.

**Comment by wtgodbe** [comment]
> F:\workspace\_work\1\s\src\ProjectTemplates\test\ProjectTemplates.Tests.csproj : error NU1201: Project Microsoft.AspNetCore.Server.IntegrationTesting is not compatible with net50 (.NETFramework,Version=v5.0). Project Microsoft.AspNetCore.Server.IntegrationTesting supports: netcoreapp5.0 (.NETCoreApp,Version=v5.0) [C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets]

Looking

**Comment by wtgodbe** [comment]
Would be because we reference that project directly, and it targets `netcoreapp5.0`:

> https://github.com/dotnet/aspnetcore/blob/26020edc02423091b2d10b2e07740255689bf3a6/src/ProjectTemplates/test/ProjectTemplates.Tests.csproj#L47

Can we get away with multitargeting either the projectTemplate test project, or `Microsoft.AspNetCore.Server.IntegrationTesting`?

**Comment by wtgodbe** [comment]
@dougbu am I reading this thread correctly, that the issue I mentioned in the comment above was supposed to be fixed by updating the machine pool to one w/ a VS new enough to have a Nuget that knows about `net50`?

**Comment by Pilchie** [comment]
> fixed by updating the machine pool to one w/ a VS new enough to have a Nuget that knows about net50?

Actually, an SDK with a new enough NuGet, but that SDK required a newer desktop msbuild (which comes with VS).  The updated machine pool that was rolled out yesterday was supposed to unblock this.

**Comment by wtgodbe** [comment]
> The updated machine pool that was rolled out yesterday was supposed to unblock this.

Was there a thread/issue about that pool? Want to make sure we're using the right one(s) in this PR

**Comment by wtgodbe** [comment]
Hmm, we are definitely using the machines that have been updated to VS 16.5.1:

https://github.com/dotnet/aspnetcore/blob/26020edc02423091b2d10b2e07740255689bf3a6/.azure/pipelines/jobs/default-build.yml#L95-L107

**Comment by Pilchie** [comment]
@markwilkie @jaredpar any ideas?

**Comment by MattGal** [comment]
> @markwilkie @jaredpar any ideas?

I'll take a look too just in case those guys are busy

**Comment by markwilkie** [comment]
Sorry, beyond confirming that the following queues have been updated to 16.5.1 on Monday, I don't have much insight.

BuildPool.Server.Amd64.VS2019.BT.Open
BuildPool.Server.Amd64.VS2019.Open
BuildPool.Server.Amd64.VS2019
Windows.Server.Amd64.VS2019.BT.Open
BuildPool.Server.Amd64.VS2019.BT

**Comment by wtgodbe** [comment]
Do we know what version of `Nuget.Client` first knew about the `net50` TFM? I'm trying to confirm that the SDK we're using actually does have a new enough Nuget

**Comment by MattGal** [comment]
I confirmed the builds are running on agents using the `BuildPool.Server.Amd64.VS2019.BT.Open/versions/2020.0403.135905` image; whatever version came with 16.5.1 buildtools is on the machine.  I'll poke about a bit more but I think this may just be expecting VS 16.5.1 to have something it might not have in box.

**Comment by wtgodbe** [comment]
Newer SDK didn't help. I don't know exactly what the requirements are here - a specific version of Nuget/Msbuild/something else on the machine/in the SDK? It appears some combination of those things is required to allow us to reference a `netcoreapp5.0` project from a `net50` project - @jaredpar do you know?

**Comment by Pilchie** [comment]
Also tagging @ericstj.

**Comment by ericstj** [comment]
So I believe this failure is in nuget restore.  From the logs it appears to be using desktop MSBuild from VS.  I believe you need to update the version of VS on the build machine to have a new nuget that understands the TFM. Has that been done?

**Comment by dougbu** [comment]
>  I believe you need to update the version of VS on the build machine to have a new nuget that understands the TFM. Has that been done?

Well, we thought so&hellip; 😈 

I would add that from what @wtgodbe and @MattGal are seeing / saying, VS 16.5.1 isn't new enough ☹️

**Comment by ericstj** [comment]
You need the VS that has our preview2 bits inserted i think.

**Comment by Pilchie** [comment]
We don’t insert .NET 5 bits into VS yet, but I think that 16.5 has a new MSBuild and NuGet which *does* understand the new TFM.

—
Kevin Pilch
https://twitter.com/Pilchie
________________________________

**Comment by dougbu** [comment]
I _think_ our main option here is to confirm 6.5.1 has an older `msbuild` or `nuget` than we need then switch to BuildPool.Server.Amd64.VS2019.Pre queues and similar.

@MattGal is there an updated list of queues available somewhere? If not, are these the correct names of preview queues that are similar to what we tend to use?
- BuildPool.Server.Amd64.VS2019.BT.Pre
- BuildPool.Server.Amd64.VS2019.BT.Pre.Open
- BuildPool.Server.Amd64.VS2019.Pre
- BuildPool.Server.Amd64.VS2019.Pre.Open

**Comment by MattGal** [comment]
> I _think_ our main option here is to confirm 6.5.1 has an older `msbuild` or `nuget` than we need then switch to BuildPool.Server.Amd64.VS2019.Pre queues and similar.
> 
> @MattGal is there an updated list of queues available somewhere? If not, are these the correct names of preview queues that are similar to what we tend to use?
> 
> * BuildPool.Server.Amd64.VS2019.BT.Pre
> * BuildPool.Server.Amd64.VS2019.BT.Pre.Open
> * BuildPool.Server.Amd64.VS2019.Pre
> * BuildPool.Server.Amd64.VS2019.Pre.Open

Some notes:
- We've never had BT.Pre queues, but the latter two do exist.
- As a "source of truth", especially in these core-reduction times, https://helix.dot.net/api/2019-06-17/info/queues will always give a (huge, JSON formatted) accurate list of what's actually available
- A less accurate (by virtue of just being an MD file), but relatively recently updated version describing what's available can be seen here: https://github.com/dotnet/arcade/blob/8ca46105193bd25c95af49bc6cd3604aaefec980/Documentation/AzureDevOps/AzureDevOpsOnboarding.md
- I'm not sure why (will have to consult with @ChulHul ) but BuildPool.Server.Amd64.VS2019.Pre is still on 16.5.0-pre.2.0+29721.120 so is significantly older than 16.5.1 RTW.

**Comment by wtgodbe** [comment]
To reiterate, these PR builds are running on `BuildPool.Server.Amd64.VS2019.BT.Open` and `BuildPool.Server.Amd64.VS2019.Open`

**Comment by dougbu** [comment]
Correct @wtgodbe and those queues use VS 16.5.1. Unfortunately, if we switched to use BuildPool.Server.Amd64.VS2019.Pre and BuildPool.Server.Amd64.VS2019.Pre.Open for all Windows jobs i.e. ignored BuildTools options, we'd be moving to a version of VS from February. Unless there's a hack we haven't tried that works around the `msbuild` / `nuget` gaps, we need newer .Pre agents.

**Comment by markwilkie** [comment]
We're hoping to get the .pre queues updated with latest by a week from Wednesday. (the 22nd)  From then on, we'll be keeping more up to date, as well as providing a scouting queue.

**Comment by Pilchie** [comment]
Why closed?

**Comment by dougbu** [comment]
Sorry I missed your question @Pilchie. Closing this was a side-effect of removing the underlying (dead) preview2 branch. I'll move it to target 'master' soon.

**Comment by Pilchie** [comment]
Let's rather make it Preview3 as that's what we're targeting for the change

**Comment by dougbu** [comment]
See #20748 for the future of this work

**Pull request closed without merging**